### PR TITLE
[SPARK-13603] [SQL] support SQL generation for subquery

### PR DESCRIPTION
**Pull request by davies**
## What changes were proposed in this pull request?

This is support SQL generation for subquery expressions, which will be replaced to a SubqueryHolder inside SQLBuilder recursively.
## How was this patch tested?

Added unit tests.

### CODE REVIEW

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/ExpressionToSQLSuite.scala [hunk 1]
```diff
@@ -268,4 +268,9 @@ class ExpressionToSQLSuite extends SQLBuilderTest with SQLTestUtils {
     checkSqlGeneration("SELECT input_file_name()")
     checkSqlGeneration("SELECT monotonically_increasing_id()")
   }
+
+  test("subquery") {
+    checkSqlGeneration("SELECT 1 + (SELECT 2)")
+    checkSqlGeneration("SELECT 1 + (SELECT 2 + (SELECT 3 as a))")
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/ExpressionToSQLSuite.scala]: Could we add more test cases?

```SELECT * from (SELECT 1 + (SELECT 2 + (SELECT 3 as a)) t``` 

If we run this, we hit a bug in Scalar Subquery support. If you want, I can try to fix it.

Comment (davies) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/ExpressionToSQLSuite.scala]: I will look into it, thanks!

Comment (davies) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/ExpressionToSQLSuite.scala]: You missed a `)` here.

I don't think this is a good test case for this suite.

**Comment by davies** [comment]
cc @liancheng

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/ExpressionToSQLSuite.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/ExpressionToSQLSuite.scala]: Sorry, my fault. : ) 

How about moving a few typical scalar queries in `SubquerySuite` to this test suite?

Comment (davies) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/ExpressionToSQLSuite.scala]: Not necessary, different test suite for different purpose, we should not repeat every test cases everywhere.

**Comment by liancheng** [comment]
LGTM, merged to master. Thanks!

**Pull request closed without merging**